Fix SQLite FK handling in altered tables

santysisi opened the pull request:
Fixes #7123

#### Summary

Fixes incorrect handling of foreign keys in `SQLitePlatform` when altering tables.

Previously, `getForeignKeys` relied on constraint names as array keys, but this behavior has changed and now returns a list.

Solution:
- Added a `$keysByName` to link constraint names to their positions
- Used this `$keysByName` to correctly unset matching foreign keys by name

This ensures that only the appropriate constraints are removed during table alterations.

santysisi committed:
Fix SQLite FK handling in altered tables

SQLitePlatform::getForeignKeysInAlteredTable was broken due to
getForeignKeys no longer returning constraint names as array keys.

This change introduces a $keysByName map to track constraints with
names, allowing dropped foreign keys to be properly matched and
removed from the result set.

santysisi commented:
Hi @morozov, I didn’t add a dedicated test in this PR because the following PR (which changes the deprecation into an error) effectively covers this new behavior. Without the fix introduced here, that next PR would fail, so it indirectly serves as a test for this change.

morozov commented:
> Without the fix introduced here, that next PR would fail, so it indirectly serves as a test for this change.

Let's make one PR for both changes. Just push the second commit to the same branch. Otherwise, there's no way to validate the fix.

santysisi committed:
Replace FK overwrite deprecation with exception

Removed the deprecation warning triggered when overwriting an existing
foreign key constraint, and replaced it with a ForeignKeyAlreadyExists
exception to enforce correct behavior.

santysisi commented:
I chose to pass a `string` instead of an `UnqualifiedName` as the second parameter of `ForeignKeyAlreadyExists` because the `key` can either come from a `UnqualifiedName` or be generated dynamically via `generateNameFromObjectColumnNames`.

That said, I’m open to changing it to `UnqualifiedName` like in `ForeignKeyDoesNotExist` if you think is a better option

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -861,7 +861,23 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
         $foreignKeys = $oldTable->getForeignKeys();
-        $nameMap     = $this->getDiffColumnNameMap($diff);
+        $keysByName  = [];
+        foreach ($foreignKeys as $key => $foreignKey) {
```
Instead of introducing a new loop, `$keysByName` could be populated in the existing one.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -905,7 +921,7 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
             );
 
-            unset($foreignKeys[$constraintKey]);
+            unset($foreignKeys[$keysByName[$constraintKey]]);
```
`$keysByName[$constraintKey]` isn't guaranteed to be set (dropping a FK that doesn't exist). This should  be an exception. Let's make an it assertion for now and revisit later.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -917,7 +933,11 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
                 );
 
-                $foreignKeys[$constraintKey] = $constraint;
+                if (isset($keysByName[$constraintKey])) {
+                    $foreignKeys[$keysByName[$constraintKey]] = $constraint;
```
This looks like an error state to me. If a constraint with the same name already exists, it shouldn't be silently overwritten.

Let's make an it assertion for now that `$keysByName[$constraintKey]` is not set and revisit later.

Also, the new name and key need to be added to `$keysByName` to prevent the overwrite by the subsequent ones.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -917,7 +933,11 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
+                if (isset($keysByName[$constraintKey])) {
+                    $foreignKeys[$keysByName[$constraintKey]] = $constraint;
+                } else {
+                    $foreignKeys[] = $constraint;
```
This code doesn't need to be duplicated. The logic, at a high level, should look like the following:
```
if the constraint is named
    make sure that there's no existing constraint with the same name
    update the name-key map
end

add the new constraint
```

morozov commented on UPGRADE.md:
```diff
@@ -8,6 +8,11 @@ awareness about deprecated code.
+## BC BREAK: changes in `Table` class
+
+The `Table` class no longer accepts `foreignKeyConstraints`
+with duplicate keys.
```
This probably should be "duplicate names".

morozov commented on src/Schema/Exception/ForeignKeyAlreadyExists.php:
```diff
@@ -0,0 +1,25 @@
+
+final class ForeignKeyAlreadyExists extends LogicException implements SchemaException
+{
+    public static function new(OptionallyQualifiedName $tableName, string $foreignKeyName): self
```
`$foreignKeyName` could still be an `UnqualifiedName`. If it's auto-generated, you can convert it using `Unqualified::unquoted()`.

I have a pending PR that will not require generating array keys for unnamed constraints, so this won't be necessary. See `OptionallyUnqualifiedNamedObjectSet`.

santysisi commented:
Hi, thanks for your suggestions!
I've made the changes based on your feedback. I think they're really good.
Please let me know if there's anything else I should adjust.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -905,7 +917,11 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
             );
 
-            unset($foreignKeys[$constraintKey]);
+            if (! isset($keysByName[$constraintKey])) {
+                continue;
+            }
```
By assertion I mean `assert(isset($keysByName[$constraintKey]))`. This way, we are saying that we always expect it to be set, otherwise it should be an error.

Making it a `continue` sends the wrong message and adds code not covered by tests.

santysisi replied:
Understand! Sorry about that, that was my fault.

morozov commented on src/Platforms/SQLitePlatform.php:
```diff
@@ -917,7 +933,11 @@ private function getForeignKeysInAlteredTable(TableDiff $diff): array
                 );
 
-                $foreignKeys[$constraintKey] = $constraint;
+                if (! isset($keysByName[$constraintKey])) {
```
Similarly. this should be `assert(! isset($keysByName[$constraintKey]))`. This will allow to move `$foreignKeys[] = $constraint;` out of the `if` as suggested in https://github.com/doctrine/dbal/pull/7158#discussion_r2366262152.

morozov commented:
Thanks, @santysisi!
